newlib/libc/sys/vita/dirent.c

```diff
@@ -92,6 +94,34 @@ DIR *opendir(const char *dirname)
 	return dirp;
 }
 
+DIR *fdopendir(int fd)
+{
+	int ret;
+	DescriptorTranslation *fdmap = __vita_fd_grab(fd);
+
+	if (!fdmap) {
+		errno = EBADF;
+		return NULL;
+	}
+
+	DIR *dirp;
+
+	if ((dirp = (DIR *)malloc(sizeof(DIR))) == NULL) {
+		errno = ENOMEM;
+		
+		__vita_fd_drop(fdmap);
+		__vita_release_descriptor(fd);
```

Comment:
@d3m3vilurr so I pushed the fix for `open`...
but still have a question about using it in dirent funcs.

so in dir-related functions and in dir struct we use sce uid for fd.
but open() and close() return other/mapped fd. how do you suggest to deal with that?

Reply:
```
int fd = open(path, O_RDONLY | O_DIRECTORY);
DIR *p = fdopendir(fd);
if (!p) {
  close(fd);
  return;
}
```
i mean just problem is this function release the fd instead waiting to call the `close`
when we got a trouble at the `malloc`, just need the set the `ENOMEM`.

`closedir` could have problem if we use it like..
```
int fd = open(path, O_RDONLY | O_DIRECTORY);
DIR *p = fdopendir(fd);
if (!p) {
  close(fd);
  return;
}
closedir(p);
```
so, an easy solution is, pass the newlib fd to `_DIR` instead of sce_uid then to change all dir related chain includes opendir, closedir, readdir...

```
DIR *opendir(const char *dirname) {
  int fd = open(dirname, O_RDONLY | O_DIRECTORY | O_CLOEXEC);
  DIR *dirp = calloc(1, sizeof(DIR);
  if (dirp == null) ...
  dirp->fd = fd;
  ...
  return drip;
}

int closedir(DIR *dirp) {
  int ret = close(dirp->fd);
  ...
  returnr et;
}

DIR *fdopendir(int fd) {
  ...
}

struct dirent *readdir(DIR *dirp) {
  ...
  DescriptorTranslation *fdmap = __vita_fd_grab(dirp->fd);
  ...
  int res = sceIoDread(fdmap->sce_uid, ....);
  ...
  __vita_fd_drop(fdmap);
  ...
}
```